File: nodejs/tests/permission_expression.test.js
// Copyright 2026, Pulumi Corporation.  All rights reserved.

// Round-trip tests for the `PermissionExpression` polymorphic hierarchy.
// Mirrors sdk/python/tests/test_permission_expression_roundtrip.py.
//
// `PermissionExpression` is the most complex generated shape: a recursive,
// discriminated composite whose fields are typed against *abstract*
// intermediates (`PermissionBooleanExpression`, `PermissionContextExpression`,
// `PermissionLiteralExpression`) while the concrete node type is carried on the
// wire in the `__type` discriminator. These tests build a deep heterogeneous
// tree, serialize it to JSON and back, and assert the reconstructed tree is
// structurally identical — exercising discriminator selection, recursion,
// abstract-field resolution, and primitive value fidelity.

const { test } = require("node:test");
const assert = require("node:assert/strict");

const { Models } = require("../dist/index.js");

// A deep, heterogeneous PermissionExpression tree (~6 levels), matching the
// Python `build_tree()` helper node-for-node.
function buildTree() {
    return Models.PermissionExpressionAnd.newInstance({
        left: Models.PermissionExpressionEqual.newInstance({
            left: Models.PermissionLiteralExpressionString.newInstance({ value: "prod" }),
            right: Models.PermissionLiteralExpressionString.newInstance({ value: "prod" }),
        }),
        right: Models.PermissionExpressionOr.newInstance({
            left: Models.PermissionExpressionNot.newInstance({
                node: Models.PermissionExpressionHasTag.newInstance({
                    node: Models.PermissionExpressionEqual.newInstance({
                        left: Models.PermissionLiteralExpressionNumber.newInstance({ value: 42.5 }),
                        right: Models.PermissionLiteralExpressionNumber.newInstance({ value: 42.5 }),
                    }),
                    context: Models.PermissionExpressionTag.newInstance({
                        context: Models.PermissionExpressionTeam.newInstance({}),
                        key: "owner",
                    }),
                    key: "team",
                }),
            }),
            right: Models.PermissionExpressionEqual.newInstance({
                left: Models.PermissionLiteralExpressionBool.newInstance({ value: true }),
                right: Models.PermissionLiteralExpressionBool.newInstance({ value: false }),
            }),
        }),
    });
}

// Serialize -> JSON string -> parse -> deserialize, as over the network.
function wireRoundtrip(obj, baseType) {
    const payload = JSON.stringify(obj);
    return Models[baseType].newInstance(JSON.parse(payload));
}

test("deep tree round-trips to a structurally identical tree", () => {
    const original = buildTree();
    const restored = wireRoundtrip(original, "PermissionExpression");

    // Deep structural equality, including reconstructed prototypes.
    assert.deepStrictEqual(restored, original);
    // The root deserialized to the correct concrete subtype.
    assert.ok(restored instanceof Models.PermissionExpressionAnd);
});

test("concrete types are selected at each level of the tree", () => {
    const restored = wireRoundtrip(buildTree(), "PermissionExpression");

    assert.ok(restored.left instanceof Models.PermissionExpressionEqual);
    assert.ok(restored.left.left instanceof Models.PermissionLiteralExpressionString);
    assert.ok(restored.right instanceof Models.PermissionExpressionOr);

    const notNode = restored.right.left;
    assert.ok(notNode instanceof Models.PermissionExpressionNot);
    const hasTag = notNode.node;
    assert.ok(hasTag instanceof Models.PermissionExpressionHasTag);
    assert.ok(hasTag.node instanceof Models.PermissionExpressionEqual);
    assert.ok(hasTag.context instanceof Models.PermissionExpressionTag);
    assert.ok(hasTag.context.context instanceof Models.PermissionExpressionTeam);
    assert.equal(hasTag.key, "team");
    assert.equal(hasTag.context.key, "owner");
});

test("wire shape carries the discriminator and primitive values", () => {
    const wire = JSON.parse(JSON.stringify(buildTree()));

    assert.equal(wire.__type, "PermissionExpressionAnd");
    assert.equal(wire.left.__type, "PermissionExpressionEqual");
    assert.equal(wire.left.left.__type, "PermissionLiteralExpressionString");
    assert.equal(wire.left.left.value, "prod");
    assert.equal(wire.right.__type, "PermissionExpressionOr");
    assert.equal(wire.right.left.node.context.context.__type, "PermissionExpressionTeam");
});

test("primitive value types are preserved through the round-trip", () => {
    const restored = wireRoundtrip(buildTree(), "PermissionExpression");

    const stringLit = restored.left.left;
    const numberLit = restored.right.left.node.node.left;
    const boolLitTrue = restored.right.right.left;
    const boolLitFalse = restored.right.right.right;

    assert.equal(typeof stringLit.value, "string");
    assert.equal(stringLit.value, "prod");

    assert.equal(typeof numberLit.value, "number");
    assert.equal(numberLit.value, 42.5);

    assert.equal(typeof boolLitTrue.value, "boolean");
    assert.equal(boolLitTrue.value, true);
    assert.equal(boolLitFalse.value, false);
});

test("deserializing through an abstract intermediate resolves the subtype", () => {
    // A tree deserialized through the abstract `PermissionBooleanExpression`
    // base must still resolve to the concrete subtype via the inherited
    // discriminator at every level.
    const original = Models.PermissionExpressionOr.newInstance({
        left: Models.PermissionExpressionEqual.newInstance({
            left: Models.PermissionLiteralExpressionString.newInstance({ value: "a" }),
            right: Models.PermissionLiteralExpressionString.newInstance({ value: "b" }),
        }),
        right: Models.PermissionExpressionNot.newInstance({
            node: Models.PermissionExpressionHasTag.newInstance({ key: "k" }),
        }),
    });
    const restored = wireRoundtrip(original, "PermissionBooleanExpression");

    assert.ok(restored instanceof Models.PermissionExpressionOr);
    assert.deepStrictEqual(restored, original);
});

test("a heterogeneous list of expressions round-trips", () => {
    const items = [
        Models.PermissionLiteralExpressionString.newInstance({ value: "s" }),
        Models.PermissionLiteralExpressionNumber.newInstance({ value: 7 }),
        Models.PermissionExpressionTeam.newInstance({}),
        buildTree(),
    ];
    const payload = JSON.stringify(items);
    const restored = JSON.parse(payload).map((i) => Models.PermissionExpression.newInstance(i));

    assert.equal(restored.length, 4);
    assert.ok(restored[0] instanceof Models.PermissionLiteralExpressionString);
    assert.ok(restored[1] instanceof Models.PermissionLiteralExpressionNumber);
    assert.ok(restored[2] instanceof Models.PermissionExpressionTeam);
    assert.ok(restored[3] instanceof Models.PermissionExpressionAnd);
    assert.deepStrictEqual(restored, items);
});

test("an empty leaf node serializes to just its discriminator", () => {
    // A concrete node with no own fields still carries its discriminator.
    const original = Models.PermissionExpressionTeam.newInstance({});
    const wire = JSON.parse(JSON.stringify(original));
    assert.deepStrictEqual(wire, { __type: "PermissionExpressionTeam" });

    const restored = Models.PermissionExpression.newInstance(wire);
    assert.ok(restored instanceof Models.PermissionExpressionTeam);
    assert.deepStrictEqual(restored, original);
});

test("an unknown discriminator falls back to the base type without throwing", () => {
    // Unlike the Python encoder (which raises on an unknown `__type`), the
    // generated TypeScript `setPrototype` switch has no default case, so an
    // unrecognized discriminator is left as the abstract base rather than
    // rejected. Pin that behavior so a future change is a conscious one.
    let restored;
    assert.doesNotThrow(() => {
        restored = Models.PermissionExpression.newInstance({ __type: "NoSuchExpression" });
    });
    assert.equal(restored.getDiscriminator(), "NoSuchExpression");
    assert.equal(Object.getPrototypeOf(restored), Models.PermissionExpression.prototype);
});

File: go/tests/permission_expression_test.go
// Copyright 2026, Pulumi Corporation.  All rights reserved.

// Round-trip tests for the `PermissionExpression` polymorphic hierarchy. Ports
// sdk/java/src/test/java/com/pulumi/cloud/sdk/PermissionExpressionRoundTripTest.java,
// which in turn mirrors sdk/python/tests/test_permission_expression_roundtrip.py
// and sdk/nodejs/tests/permission_expression.test.js. The tree built here matches
// those `buildTree()` helpers node-for-node.
//
// `PermissionExpression` is the most complex generated shape: a recursive,
// discriminated composite whose fields are typed against *abstract* intermediates
// (PermissionBooleanExpression) while the concrete node type travels in the
// `__type` discriminator.
//
// Go models this differently from Java/C#: an exported sealed interface per node,
// an unexported impl, an exported `…Builder{}.Build()` constructor, and a
// free-function dispatcher (`UnmarshalJSONPermissionExpression`) rather than an
// UnmarshalJSON on the interface. Tests therefore construct via builders and
// assert via interface type assertions — the impls are unexported and
// unreachable.
package tests

import (
	"encoding/json"
	"strings"
	"testing"

	"github.com/pulumi/pulumi-cloud-sdk/go/apitype"
)

func str(v string) apitype.PermissionLiteralExpressionString {
	return apitype.PermissionLiteralExpressionStringBuilder{Value: v}.Build()
}

func num(v float64) apitype.PermissionLiteralExpressionNumber {
	return apitype.PermissionLiteralExpressionNumberBuilder{Value: v}.Build()
}

func boolean(v bool) apitype.PermissionLiteralExpressionBool {
	return apitype.PermissionLiteralExpressionBoolBuilder{Value: v}.Build()
}

func eq(left, right apitype.PermissionExpression) apitype.PermissionExpressionEqual {
	return apitype.PermissionExpressionEqualBuilder{Left: left, Right: right}.Build()
}

// A deep, heterogeneous PermissionExpression tree:
//
//	And( Equal("prod", "prod"), Or( Equal(42.5, 42.5), Equal(true, false) ) )
func buildTree() apitype.PermissionExpressionAnd {
	or := apitype.PermissionExpressionOrBuilder{
		PermissionBooleanExpressionBinaryBuilder: apitype.PermissionBooleanExpressionBinaryBuilder{
			Left:  eq(num(42.5), num(42.5)),
			Right: eq(boolean(true), boolean(false)),
		},
	}.Build()

	return apitype.PermissionExpressionAndBuilder{
		PermissionBooleanExpressionBinaryBuilder: apitype.PermissionBooleanExpressionBinaryBuilder{
			Left:  eq(str("prod"), str("prod")),
			Right: or,
		},
	}.Build()
}

func TestRoundTripPreservesConcreteTypesAndValues(t *testing.T) {
	t.Parallel()

	encoded, err := json.Marshal(buildTree())
	if err != nil {
		t.Fatalf("marshalling tree: %v", err)
	}

	// The discriminator must be on the wire for every concrete node, otherwise
	// the dispatcher below has nothing to switch on.
	for _, want := range []string{
		`"__type":"PermissionExpressionAnd"`,
		`"__type":"PermissionExpressionOr"`,
		`"__type":"PermissionExpressionEqual"`,
		`"__type":"PermissionLiteralExpressionString"`,
		`"__type":"PermissionLiteralExpressionNumber"`,
		`"__type":"PermissionLiteralExpressionBool"`,
	} {
		if !strings.Contains(string(encoded), want) {
			t.Errorf("serialized tree is missing %s\ngot: %s", want, encoded)
		}
	}

	// Re-parse as the *abstract* root and walk back down, asserting the concrete
	// type at every level.
	var decoded apitype.PermissionExpression
	if err := apitype.UnmarshalJSONPermissionExpression(encoded, &decoded); err != nil {
		t.Fatalf("unmarshalling tree: %v", err)
	}

	and, ok := decoded.(apitype.PermissionExpressionAnd)
	if !ok {
		t.Fatalf("root: got %T, want PermissionExpressionAnd", decoded)
	}

	andLeft, ok := and.Left().(apitype.PermissionExpressionEqual)
	if !ok {
		t.Fatalf("and.Left: got %T, want PermissionExpressionEqual", and.Left())
	}
	assertString(t, "and.Left.Left", andLeft.Left(), "prod")
	assertString(t, "and.Left.Right", andLeft.Right(), "prod")

	or, ok := and.Right().(apitype.PermissionExpressionOr)
	if !ok {
		t.Fatalf("and.Right: got %T, want PermissionExpressionOr", and.Right())
	}

	orLeft, ok := or.Left().(apitype.PermissionExpressionEqual)
	if !ok {
		t.Fatalf("or.Left: got %T, want PermissionExpressionEqual", or.Left())
	}
	assertNumber(t, "or.Left.Left", orLeft.Left(), 42.5)
	assertNumber(t, "or.Left.Right", orLeft.Right(), 42.5)

	orRight, ok := or.Right().(apitype.PermissionExpressionEqual)
	if !ok {
		t.Fatalf("or.Right: got %T, want PermissionExpressionEqual", or.Right())
	}
	assertBool(t, "or.Right.Left", orRight.Left(), true)
	assertBool(t, "or.Right.Right", orRight.Right(), false)
}

// Java's `unknownDiscriminatorFails`, which expects Jackson's
// InvalidTypeIdException. The Go dispatcher's default branch reports
// "type '%s' not recognized".
func TestUnknownDiscriminatorFails(t *testing.T) {
	t.Parallel()

	var decoded apitype.PermissionExpression
	err := apitype.UnmarshalJSONPermissionExpression([]byte(`{"__type":"NoSuchExpression"}`), &decoded)
	if err == nil {
		t.Fatal("expected an error for an unknown discriminator, got nil")
	}
	if !strings.Contains(err.Error(), "NoSuchExpression") {
		t.Errorf("error should name the offending discriminator, got: %v", err)
	}
}

// The abstract root cannot be marshalled — it has no discriminator to emit. This
// has no Java or C# counterpart; it is specific to Go's builder-per-interface
// modelling, where the abstract builder is constructible.
func TestAbstractExpressionCannotBeMarshalled(t *testing.T) {
	t.Parallel()

	// Parenthesised: a composite literal in an if-statement initialiser is
	// ambiguous with the block that follows it.
	abstract := (apitype.PermissionExpressionBuilder{}).Build()
	if _, err := abstract.MarshalJSON(); err == nil {
		t.Fatal("expected marshalling the abstract PermissionExpression to fail, got nil")
	}
}

func assertString(t *testing.T, field string, got apitype.PermissionExpression, want string) {
	t.Helper()
	lit, ok := got.(apitype.PermissionLiteralExpressionString)
	if !ok {
		t.Fatalf("%s: got %T, want PermissionLiteralExpressionString", field, got)
	}
	if lit.Value() != want {
		t.Errorf("%s: got %q, want %q", field, lit.Value(), want)
	}
}

func assertNumber(t *testing.T, field string, got apitype.PermissionExpression, want float64) {
	t.Helper()
	lit, ok := got.(apitype.PermissionLiteralExpressionNumber)
	if !ok {
		t.Fatalf("%s: got %T, want PermissionLiteralExpressionNumber", field, got)
	}
	if lit.Value() != want {
		t.Errorf("%s: got %v, want %v", field, lit.Value(), want)
	}
}

func assertBool(t *testing.T, field string, got apitype.PermissionExpression, want bool) {
	t.Helper()
	lit, ok := got.(apitype.PermissionLiteralExpressionBool)
	if !ok {
		t.Fatalf("%s: got %T, want PermissionLiteralExpressionBool", field, got)
	}
	if lit.Value() != want {
		t.Errorf("%s: got %v, want %v", field, lit.Value(), want)
	}
}
